base_layer/core/src/base_node/service/blockchain_state/service.rs

```diff
@@ -0,0 +1,211 @@
+    StreamExt,
+    TryFutureExt,
+};
+use log::*;
+use std::{future::Future, ops::Bound};
+use tari_comms::protocol::rpc::__macro_reexports::BoxFuture;
+use tokio::future::poll_fn;
+
+const LOG_TARGET: &str = "c::base_node::blockchain_state_service";
+
+pub struct BlockchainStateService<B> {
+    db: BlockchainDatabase<B>,
+    requests: Fuse<mpsc::Receiver<BlockchainStateRequest>>,
+}
+
+impl<B: BlockchainBackend + 'static> BlockchainStateService<B> {
+    pub fn new(db: BlockchainDatabase<B>, requests: mpsc::Receiver<BlockchainStateRequest>) -> Self {
+        Self {
+            db,
+            requests: requests.fuse(),
+        }
+    }
+
+    #[inline]
+    fn db(&self) -> BlockchainDatabase<B> {
+        self.db.clone()
+    }
+
+    pub async fn run(mut self) {
+        // This "work queue" holds all the pending tasks. The reason this is a good fit for this service is that this
+        // service is only for calls to the database wrapped in some ergonomics. All async_db calls spawn a
+        // blocking task. So the work queue doesn't actually have to do much work itself, it is simply waiting
+        // for the blocking task to notify it when it is complete, so polling many of these tasks sequentially is very
+        // lightweight.
+        let mut work_queue = FuturesUnordered::new();
+
+        loop {
+            futures::select! {
+                // Add each request to a "work queue"
+                request = self.requests.select_next_some() => {
+                    let work = self.handle_request(request);
+                    work_queue.push(work);
+                },
+                // Process the work queue
+                _ = work_queue.select_next_some() => { },
+            }
+        }
+    }
+
+    fn handle_request(&mut self, request: BlockchainStateRequest) -> BoxFuture<'static, ()> {
+        use BlockchainStateRequest::*;
+        trace!(target: LOG_TARGET, "Got request: {:?}", request);
+        match request {
+            GetBlocks((start, end), reply) => {
+                Box::pin(reply_or_cancel(reply, Self::fetch_blocks(self.db(), start, end)))
+            },
+            GetHeaders((start, end), reply) => {
+                Box::pin(reply_or_cancel(reply, Self::fetch_headers(self.db(), start, end)))
+            },
+            GetHeaderByHeight(height, reply) => Box::pin(reply_or_cancel(
+                reply,
+                async_db::fetch_header(self.db(), height).map_err(Into::into),
+            )),
+            GetChainMetadata(reply) => Box::pin(reply_or_cancel(
+                reply,
+                async_db::get_chain_metadata(self.db()).map_err(Into::into),
+            )),
+            FindFirstMatchingHeader(hashes, reply) => Box::pin(reply_or_cancel(
+                reply,
+                Self::find_first_matching_header(self.db(), hashes),
+            )),
+        }
+    }
+
+    async fn fetch_blocks(
+        db: BlockchainDatabase<B>,
+        start: Bound<u64>,
```

Comment:
Apart from local querying and GRPC, I don't see the value in using heights instead of hashes. The sync node's chain could change during this step or in between calls, making this query invalid. Perhaps pass in a starting hash and a number of blocks.

Reply:
Great idea! Will change the api.